Data/ByteString/Internal.hs

```diff
@@ -339,18 +339,18 @@ packChars cs = unsafePackLenChars (List.length cs) cs
  #-}
 
 unsafePackLenBytes :: Int -> [Word8] -> ByteString
-unsafePackLenBytes len xs0 =
-    unsafeCreate len $ \p -> go p xs0
-  where
-    go !_ []     = return ()
-    go !p (x:xs) = poke p x >> go (p `plusPtr` 1) xs
+unsafePackLenBytes len =
+    unsafeCreate len . foldr
+      (\x go p -> poke p x >> go (p `plusPtr` 1))
+      (\_ -> return ())
+{-# INLINE unsafePackLenBytes #-}
```

Comment:
I wonder why these functions have not been inlined before.

Reply:
As the benchmarks show, if no fusion happens then there is no benefit in inlining (it even costs a few percents of performance).

Reply:
I now think this is actually quite a big problem, because the only way this function is exposed in the safe API is through `pack` which can never fuse anyway, so is this patch really still desirable? We would need a safe `packLen` function to really get the benefits.

Reply:
Here's a version of `pack` that is fusable (by using dynamic arrays doubling in size):

https://gist.github.com/noughtmare/8902ddec65fedae9fb24d3826dbe11a0

This again has the problem that it is slightly slower if not fused, but a lot faster if it is fused.